iroh-bytes/src/baomap.rs

```diff
@@ -0,0 +1,288 @@
+/// to copy small files even if the mode is `Reference`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub enum ImportMode {
+    /// This mode will copy the file into the database before hashing.
+    ///
+    /// This is the safe default because the file can not be accidentally modified
+    /// after it has been imported.
+    #[default]
+    Copy,
+    /// This mode will reference the file in place and assume it is unchanged after import.
+    ///
+    /// This has a large performance and storage benefit, but it is less safe since
+    /// the file might be modified after it has been imported.
+    Reference,
+}
+/// The import mode describes how files will be imported.
+///
+/// This is a hint to the import trait method. For some implementations, this
+/// does not make any sense. E.g. an in memory implementation will always have
+/// to copy the file into memory. Also, a disk based implementation might choose
+/// to copy small files even if the mode is `Reference`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub enum ExportMode {
+    /// This mode will copy the file to the target directory.
+    ///
+    /// This is the safe default because the file can not be accidentally modified
+    /// after it has been exported.
+    #[default]
+    Copy,
+    /// This mode will move the file to the target directory and then reference it from
+    /// the database.
+    ///
+    /// This has a large performance and storage benefit, but it is less safe since
+    /// the file might be modified in the target directory after it has been exported.
+    Reference,
+}
+
+#[allow(missing_docs)]
+#[derive(Debug)]
+pub enum ExportProgress {
+    /// Starting to export to a file
+    ///
+    /// This will be the first message for an id
+    Start {
+        id: u64,
+        hash: Hash,
+        path: PathBuf,
+        stable: bool,
+    },
+    /// Progress when copying the file to the target
+    ///
+    /// This will be omitted if the store can move the file or use copy on write
+    ///
+    /// There will be multiple of these messages for an id
+    Progress { id: u64, offset: u64 },
+    /// Done exporting
+    Done { id: u64 },
+}
+
+/// Progress updates for the provide operation
+#[derive(Debug, Serialize, Deserialize)]
+pub enum ValidateProgress {
+    /// started validating
+    Starting {
+        /// The total number of entries to validate
+        total: u64,
+    },
+    /// We started validating an entry
+    Entry {
+        /// a new unique id for this entry
+        id: u64,
+        /// the hash of the entry
+        hash: Hash,
+        /// location of the entry.
+        ///
+        /// In case of a file, this is the path to the file.
+        /// Otherwise it might be an url or something else to uniquely identify the entry.
+        path: Option<String>,
+        /// the size of the entry
+        size: u64,
+    },
+    /// We got progress ingesting item `id`.
+    Progress {
```

Comment:
so this happens many times for each `Entry`? maybe then the names would be a little better as 

- `EntryStart`
- `EntryProgress`
- `EntryDone`

Reply:
In some cases there can be many different progress types. E.g. if you import you might want to have progress for the outboard computation / hashing as well as for the copying into the store.

But `EntryCopyStart`, `EntryCopyProgress`, `EntryCopyDone` would sound a bit long. I thought that the fact that the msg has an id is sufficient to make it clear that it refers to an entry.

But I am not wed to the names.

Reply:
Yeah, they do get long. Maybe just add sth to the comment in that regard, that it gets emitted once or multiple times per X